### DISCUSSION TITLE
MIDI pipe "exclusive"

### DISCUSSION THREAD

**Discussion by fitzgreyve** [tttapa/Control-Surface#910] [General]
Pieter, great library.

I have defined three pipes (two are to replace the default Control_Surface ones):
MIDI_Pipe txPipe;                     
MIDI_Pipe rxPipe;                     
MIDI_Pipe thruPipe;  

then:
  Control_Surface >> txPipe >> serialMIDI;  // MIDI output pipe.
  Control_Surface << rxPipe << serialMIDI;  // MIDI input pipe.
  serialMIDI >> thruPipe >> serialMIDI;     // loopback all incoming MIDI to the output.

That all works OK.

I need txPipe to have exlusive use of the serialMIDI output while sending NPRN's, but can't get the syntax of the "exclusive" function. I had assumed:
txPipe.exclusive(0, true);    but that gives  'txPipe' is not a class, namespace, or enumeration

Can you clarify how to correctly use "exclusive"  please, I did find a reference to it, but it is currently being elusive.

Many thanks.

**Comment by tttapa** [comment]
Hi,

Please post a complete example with the complete error message.

If you're sending NRPN using Control Surface, exclusive access is not necessary (except if you call `Control_Surface.update()` in between the `Control_Surface.sendControlChange()` calls that send a single NRPN message, which is probably a bad idea either way).

If you need to receive NRPN from the `serialMIDI` input, that would require some changes to the `MIDI_Interface::updateIncoming` function:

https://github.com/tttapa/Control-Surface/blob/39182bbab79487236c11579db3acfa336813bf54/src/MIDI_Interfaces/MIDI_Interface.hpp#L121-L140

You can see how the exclusive access works: if you call `source.stall(staller)`, that `source` gets exclusive access to all pipes and sinks that it can send data to. When another source tries to send data to one of those pipes, it has to wait, and it calls back the `staller.handleStall(source)` function, which should un-stall the `source`. Once the original source is un-stalled, the other source can send its data.

**Comment by fitzgreyve** [comment]
Thanks. 
This is outgoing MIDI only, and there are no Control_Surface.update() during each NRPN message sequence. 
I'll ignore the exclusive access for now.
